pkg/objstore/backoff.go

```diff
@@ -0,0 +1,248 @@
+			// Retry again after backoff
+		}
+
+		r, err = b.Bucket.GetRange(ctx, name, off, length)
+		// Don't retry in case of 404 or if it's an object worth ignoring.
+		if err == nil || b.Bucket.IsObjNotFoundErr(err) || b.isIgnoreObject(name) {
+			return r, err
+		}
+
+		level.Warn(b.logger).Log("msg", "get range failed", "name", name, "retryAttempt", retry, "error", err)
+		b.retries.WithLabelValues(op).Inc()
+	}
+
+	return r, err
+}
+
+func (b *backoffBucket) Exists(ctx context.Context, name string) (bool, error) {
+	const op = OpExists
+	exists, err := b.Bucket.Exists(ctx, name)
+	// Don't retry in case of 404 or if it's an object worth ignoring.
+	if err == nil || b.Bucket.IsObjNotFoundErr(err) || b.isIgnoreObject(name) {
+		return exists, err
+	}
+
+	for retry := 1; retry <= b.Backoff.Max(); retry++ {
+		select {
+		case <-ctx.Done():
+			return exists, err
+		case <-b.Backoff.Done():
+			return exists, err
+		case <-time.After(b.Backoff.Duration(retry)):
+			// Retry again after backoff
+		}
+
+		exists, err = b.Bucket.Exists(ctx, name)
+		// Don't retry in case of 404 or if it's an object worth ignoring.
+		if err == nil || b.Bucket.IsObjNotFoundErr(err) || b.isIgnoreObject(name) {
+			return exists, err
+		}
+
+		level.Warn(b.logger).Log("msg", "exists failed", "name", name, "retryAttempt", retry, "error", err)
+		b.retries.WithLabelValues(op).Inc()
+	}
+
+	return exists, err
+}
+
+func (b *backoffBucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	const op = OpUpload
+
+	err := b.Bucket.Upload(ctx, name, r)
+	if err == nil || b.isIgnoreObject(name) {
+		return nil
+	}
+
+	for retry := 1; retry <= b.Backoff.Max(); retry++ {
+		select {
+		case <-ctx.Done():
+			return err
+		case <-b.Backoff.Done():
+			return err
+		case <-time.After(b.Backoff.Duration(retry)):
+			// Retry again after backoff
+		}
+
+		err = b.Bucket.Upload(ctx, name, r)
+		if err == nil || b.isIgnoreObject(name) {
+			return nil
+		}
+
+		level.Warn(b.logger).Log("msg", "upload failed", "name", name, "retryAttempt", retry, "error", err)
+		b.retries.WithLabelValues(op).Inc()
+	}
+
+	return err
+}
+
+func (b *backoffBucket) Delete(ctx context.Context, name string) error {
+	const op = OpDelete
+
+	err := b.Bucket.Delete(ctx, name)
+	if err == nil || b.isIgnoreObject(name) {
+		return nil
+	}
+
+	for retry := 1; retry <= b.Backoff.Max(); retry++ {
+		select {
+		case <-ctx.Done():
+			return err
+		case <-b.Backoff.Done():
+			return err
+		case <-time.After(b.Backoff.Duration(retry)):
+			// Retry again after backoff
+		}
+
+		err = b.Bucket.Delete(ctx, name)
+		if err == nil || b.isIgnoreObject(name) {
+			return nil
+		}
+
+		level.Warn(b.logger).Log("msg", "delete failed", "name", name, "retryAttempt", retry, "error", err)
+		b.retries.WithLabelValues(op).Inc()
+	}
+
+	return err
+}
+
+func (b *backoffBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	return b.Bucket.Iter(ctx, dir, f)
```

Comment:
A problem of handling retries with a wrapper is that `Iter()` can't be retried. Despite harder to build, I personally believe retries should be handled at a lower level.